src/Analyser/MutatingScope.php

```diff
@@ -1618,21 +1618,23 @@ private function resolveType(string $exprString, Expr $node): Type
 
 			return new ErrorType();
 		} elseif ($node instanceof Node\Expr\ClassConstFetch && $node->name instanceof Node\Identifier) {
+			if ($this->isInTrait()
+				&& $node->class instanceof Name
+				&& $node->class->toString() === 'self'
```

Comment:
That's not the only way how to reference the current class ;) You need to resolve the name and then compare it against the current class name.

Also test how self::class looks like.

Also - the correct place to do this is probably InitializerExprTypeResolver.

Reply:
> That's not the only way how to reference the current class ;) You need to resolve the name and then compare it against the current class name.

generalizing `static::class` to `'class-string&literal-string'` makes some existing tests fail unexpectedly

```
1) PHPStan\Rules\Comparison\ImpossibleCheckTypeFunctionCallRuleTest::testImpossibleCheckTypeFunctionCall
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 634: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'method' will always evaluate to true.
 637: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'someAnother' will always evaluate to true.
 640: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'unknown' will always evaluate to false.
-643: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'method' will always evaluate to true.
-646: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'someAnother' will always evaluate to true.
-649: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'unknown' will always evaluate to false.
+643: Call to function method_exists() with class-string&literal-string and 'method' will always evaluate to false.
+646: Call to function method_exists() with class-string&literal-string and 'someAnother' will always evaluate to false.
+649: Call to function method_exists() with class-string&literal-string and 'unknown' will always evaluate to false.
 678: Call to function is_string() with string will always evaluate to true.
     💡 Because the type is coming from a PHPDoc, you can turn off this check by setting <fg=cyan>treatPhpDocTypesAsCertain: false</> in your <fg=cyan>%configurationFile%</>.
 693: Call to function assert() with true will always evaluate to true.

C:\dvl\Workspace\phpstan-src-staabm\src\Testing\RuleTestCase.php:153
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Rules\Comparison\ImpossibleCheckTypeFunctionCallRuleTest.php:48

2) PHPStan\Rules\Comparison\ImpossibleCheckTypeFunctionCallRuleTest::testImpossibleCheckTypeFunctionCallWithoutAlwaysTrue
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 598: Call to function method_exists() with 'UndefinedClass' and 'test' will always evaluate to false.
 631: Call to function method_exists() with $this(CheckTypeFunctionCall\MethodExistsWithTrait) and 'unknown' will always evaluate to false.
 640: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'unknown' will always evaluate to false.
-649: Call to function method_exists() with 'CheckTypeFunctionCall\\MethodExistsWithTrait' and 'unknown' will always evaluate to false.
+643: Call to function method_exists() with class-string&literal-string and 'method' will always evaluate to false.
+646: Call to function method_exists() with class-string&literal-string and 'someAnother' will always evaluate to false.
+649: Call to function method_exists() with class-string&literal-string and 'unknown' will always evaluate to false.
 694: Call to function assert() with false will always evaluate to false.
 694: Call to function is_numeric() with 'blabla' will always evaluate to false.
 '

C:\dvl\Workspace\phpstan-src-staabm\src\Testing\RuleTestCase.php:153
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Rules\Comparison\ImpossibleCheckTypeFunctionCallRuleTest.php:260

3) PHPStan\Analyser\AnalyserIntegrationTest::testBug4288
Failed asserting that PHPStan\Type\MixedType Object (...) is an instance of class "PHPStan\Type\Constant\ConstantIntegerType".

C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\AnalyserIntegrationTest.php:373
```

Reply:
Looks like a bug, you need to check MethodExistsTypeSpecifyingExtension and also there's something about `method_exists` in ImpossibleCheckTypeHelper.

Reply:
thank you.

with https://github.com/phpstan/phpstan-src/pull/2046/commits/4d6944791918ecc50ec66564ca7e43fe930c362b I have adjusted `MethodExistsTypeSpecifyingExtension` so it no longer makes the impossible type checker emit errors for `static::class` when used within traits. My reasoning is, since we generalized the class-string type when used within traits, we can no longer reason about always true/false method errors for this expression.

the remaining failing test in https://github.com/phpstan/phpstan-src/blob/e3bf63090b9da7d3e8ef8496eb8c5e7f1fa47da4/tests/PHPStan/Analyser/data/bug-4288.php#L21 is also related to the generalization. since we generalize the constant, we no longer can hold the expectations of this testcase.

I guess this is not accaptable.. but lets see what you think :)

Reply:
> Looks like a bug, you need to check MethodExistsTypeSpecifyingExtension and also there's something about `method_exists` in ImpossibleCheckTypeHelper.

extracted into a separate issue: https://github.com/phpstan/phpstan/issues/8752